### PR TITLE
test COPY and implement "correct" behavior in ch-grow

### PR DESCRIPTION
**Pull request by reidpr**
Addresses #654.

### CODE REVIEW

**Comment by reidpr** [comment]
Note that this PR currently includes an old version of branch `test-one-file` (PR #704). Once that PR is merged, I'll merge from master and the changes in this PR will be more clear.

File: bin/ch-grow.py.in [hunk 1]
```diff
@@ -191,7 +190,10 @@ environment variables:
    state = State()
 
    parser = lark.Lark(GRAMMAR, parser="earley", propagate_positions=True)
-   text = open(cli.file, "rt").read()
+   if (cli.file == "-"):
```
Comment (rstyd) [inline bin/ch-grow.py.in:193]: What motivated this change?

Comment (reidpr) [reply bin/ch-grow.py.in:193]: Good question. It's common for UNIX apps to accept `-` (single hyphen) as a filename indicating standard input or output. `docker` does this, and I added it in this PR for parity to make the tests pass.

File: test/fixtures/README [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Currently there is only one fixture and it's a symlink. This README serves as
```
Comment (rstyd) [inline test/fixtures/README:1]: Can you extrapolate on what purpose the fixtures serve?

Comment (reidpr) [reply test/fixtures/README:1]: They're files we need for testing. As-is, we don't actually store the one existing fixture because it's a symlink, which messes up Autotools, but that directory is where it goes.

File: test/build/50_dockerfile.bats [hunk 3]
```diff
@@ -197,3 +197,105 @@ EOF
...
     fi
 }
+
+@test 'Dockerfile: COPY errors' {
+    scope standard
+    [[ $CH_BUILDER = none ]] && skip 'no builder'
+    [[ $CH_BUILDER = buildah* ]] && skip 'Buildah untested'
+
+    # Dockerfile on stdin, so no context directory.
+    if [[ $CH_BUILDER != ch-grow ]]; then  # ch-grow doesn't support this yet
```
Comment (rstyd) [inline test/build/50_dockerfile.bats:207]: What does this mean RE ch-grow doesn't support this yet?

File: test/fixtures/README (hunk 2 above)
Comment (rstyd) [reply test/fixtures/README:1]: I would like to see an additional sentence here that just explains that the fixtures are for testing COPY in ch-grow.

File: bin/ch-test [hunk 4]
```diff
@@ -520,7 +520,7 @@ test_make_tag_from_path () {
     #   3) basicname:  the file name (Dockerfile)
     #   4) extension:  the file's extension (openmpi)
     #
-    # $basicname must be 'Build' or 'Dockerfile'; otherwise error.
+    # $basicname must be 'Build', 'Dockerfile'; otherwise error.
```
Comment (rstyd) [inline bin/ch-test:523]: Why was this changed to `,`? `or` seems to be the correct punctuation?

File: bin/ch-grow.py.in [hunk 5]
```diff
@@ -269,32 +271,68 @@ class Copy(Instruction):
...
+         ch.mkdirs(dst)
       for src in srcs:
-         cp(src, dst)
-
+         # Check for symlinks to outside context.
+         src_real = os.path.realpath(src)
+         context_real = os.path.realpath(cli.context)
+         if (not os.path.commonpath([src_real, context_real]) \
+                 .startswith(context_real)):
+            FATAL("can't COPY: %s uses symlinks to get outside context" % src)
```
Comment (rstyd) [inline bin/ch-grow.py.in:306]: We should consider stating that symlinks can't go outside the context in the documentation.

File: test/build/50_dockerfile.bats (hunk 3 above)
Comment (reidpr) [reply test/build/50_dockerfile.bats:207]: If you give Docker a Dockerfile on stdin without `-f`, it has no context directory. `ch-grow` doesn't support not having a context directory.

File: bin/ch-test (hunk 4 above)
Comment (reidpr) [reply bin/ch-test:523]: That's a mistake. Thanks; I thought I changed it back.

File: bin/ch-grow.py.in (hunk 5 above)
Comment (reidpr) [reply bin/ch-grow.py.in:306]: OK, I'll think about it. I believe Docker's Dockerfile documentation talks about how you can only use files within the context directory.

File: test/build/50_dockerfile.bats (hunk 3 above)
Comment (rstyd) [reply test/build/50_dockerfile.bats:207]: Ah, okay.

File: bin/ch-grow.py.in (hunk 1 above)
Comment (j-ogas) [reply bin/ch-grow.py.in:193]: Is this argument covered in the copy/test.bats file?

Comment (reidpr) [reply bin/ch-grow.py.in:193]: It's in one of the Bats files; not sure which one.
